src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -88,7 +88,10 @@ private bool EnqueueCore(TElement value)
                 waiter = _waiters.Dequeue();
             }
 
-            waiter.SetResult(value);
+            // Invoke SetResult on a separate task, as this invocation could cause the underlying task to executing,
+            // which could be a long running operation that can potentially cause a deadlock if executed on the current thread.
+            Task.Run(() => waiter.SetResult(value));
```

Comment:
RunContinuationsAsynchronously is in the System.Threading.Tasks 4.0.10 contract.  I don't believe it's available in .NET 4.5.1, unfortunately.  Is that what Roslyn is using?